File: Samples/iOS/Objective-C/IOS-SDK-Demo-App/Ad Formats/LORewardedViewController.m
//
//  LORewardedViewController.m
//  test-objc-fw-integration
//
//  Created by Clarke Bishop on 3/29/23.
//

/**
 IMPORTANT: Please review and set the placementID and appID inside the Info.plist file. To do so, please do the following:
 
 Navigate to project hierarchy in Xcode, and select Info. Expand Liftoff PlacementID and be sure that Interstitial is set correctly. Please do the same for Liftoff Dashboard AppID as well inside the Info.plist file.
 
 */

#import "LORewardedViewController.h"
#import <VungleAdsSDK/VungleAdsSDK.h>
#import "CallbackLog.h"
#import "AppUtil.h"
#import "CallbackLogTableViewCell.h"

@interface LORewardedViewController () <VungleRewardedDelegate, UITableViewDelegate, UITableViewDataSource>
@property (nonatomic, strong) VungleRewarded *rewardedAd;
@property (nonatomic, strong) NSArray *callbackLogs;

@property (nonatomic, weak) IBOutlet UILabel *placementLbl;
@property (nonatomic, weak) IBOutlet UIButton *loadBtn;
@property (nonatomic, weak) IBOutlet UIButton *playBtn;
@property (nonatomic, weak) IBOutlet UITableView *tableView;
@end

@implementation LORewardedViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  // Do any additional setup after loading the view.
  
  self.navigationItem.title = @"Liftoff Rewarded Ad";
  self.placementLbl.text = self.placementId;
  self.callbackLogs = [self generateCallbackLogs];
  
  self.tableView.delegate = self;
  self.tableView.dataSource = self;  
}

#pragma mark - IBAction Methods

- (IBAction)btnPressed:(id)sender {
  if (sender == self.loadBtn && self.placementId != nil) {
    [self loadRewardedAd];
  } else if (sender == self.playBtn && [self.rewardedAd canPlayAd]) {
    [self playRewardedAd];
  }
}

#pragma mark - Private Methods

- (void)loadRewardedAd {
  // clean up before load
  if (self.rewardedAd != nil) {
    self.rewardedAd.delegate = nil;
    self.rewardedAd = nil;
    [self resetCallbackLogs];
  }
  
  if (self.placementId) {
    self.rewardedAd = [[VungleRewarded alloc] initWithPlacementId:self.placementId];
    self.rewardedAd.delegate = self;
    [self.rewardedAd load:nil];
    
    // Set incentivized rewarded text
    [self.rewardedAd setAlertTitleText:@"Close Ad"];
    [self.rewardedAd setAlertBodyText:@"Are you sure you want to close the ad? You will lose out on your reward"];
    [self.rewardedAd setAlertCloseButtonText:@"Close Ad"];
    [self.rewardedAd setAlertContinueButtonText:@"Resume Ad"];
  }
}

- (void)playRewardedAd {
  [self.rewardedAd presentWith:self];
}

- (NSArray *)generateCallbackLogs {
  return @[
    [[CallbackLog alloc] initWithTitle:@"rewardedAdDidLoad" index:0],
    [[CallbackLog alloc] initWithTitle:@"rewardedAdWillPresent" index:1],
    [[CallbackLog alloc] initWithTitle:@"rewardedAdDidPresent" index:2],
    [[CallbackLog alloc] initWithTitle:@"rewardedAdDidTrackImpression" index:3],
    [[CallbackLog alloc] initWithTitle:@"rewardedAdDidClick" index:4],
    [[CallbackLog alloc] initWithTitle:@"rewardedAdWillLeaveApplication" index:5],
    [[CallbackLog alloc] initWithTitle:@"rewardedAdDidRewardUser" index:6],
    [[CallbackLog alloc] initWithTitle:@"rewardedAdWillClose" index:7],
    [[CallbackLog alloc] initWithTitle:@"rewardedAdDidClose" index:8],
    [[CallbackLog alloc] initWithTitle:@"rewardedAdDidFailToLoad" index:9],
    [[CallbackLog alloc] initWithTitle:@"rewardedAdDidFailToPresent" index:10]
  ];
}

- (void)updateCallbackLogsWithTitle:(NSString *)title {
  dispatch_async(dispatch_get_main_queue(), ^{
    self.callbackLogs = [AppUtil updateCallbackLogs:self.callbackLogs title:title];
    [self.tableView reloadData];
  });
}

- (void)resetCallbackLogs {
  self.callbackLogs = [AppUtil resetCallbackLogs:self.callbackLogs];
  [self.tableView reloadData];
}

#pragma mark - UITableView Datasource / Delegate Methods

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  return self.callbackLogs.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  CallbackLogTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"CallbackLogTableCell" forIndexPath:indexPath];
  CallbackLog *log = self.callbackLogs[indexPath.row];
  if (log.isActive) {
    [cell.callbackLbl setTextColor:[UIColor labelColor]];
    cell.accessoryType = UITableViewCellAccessoryCheckmark;
  } else {
    [cell.callbackLbl setTextColor:[UIColor grayColor]];
    cell.accessoryType = UITableViewCellAccessoryNone;
  }
  cell.callbackLbl.text = log.title;
  return cell;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
  return 1;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
  return 60;
}

#pragma mark - VungleRewarded Delegate Methods

- (void)rewardedAdDidLoad:(VungleRewarded *)rewarded {
  NSLog(@"rewardedAdDidLoad");
  [self updateCallbackLogsWithTitle:@"rewardedAdDidLoad"];
}

- (void)rewardedAdDidFailToLoad:(VungleRewarded *)rewarded
                      withError:(NSError *)withError {
  NSLog(@"rewardedAdDidFailToLoad");
  [self updateCallbackLogsWithTitle:@"rewardedAdDidFailToLoad"];
}

- (void)rewardedAdWillPresent:(VungleRewarded *)rewarded {
  NSLog(@"rewardedAdWillPresent");
  [self updateCallbackLogsWithTitle:@"rewardedAdWillPresent"];
}

- (void)rewardedAdDidPresent:(VungleRewarded *)rewarded {
  NSLog(@"rewardedAdDidPresent");
  [self updateCallbackLogsWithTitle:@"rewardedAdDidPresent"];
}

- (void)rewardedAdDidFailToPresent:(VungleRewarded *)rewarded
                         withError:(NSError *)withError {
  NSLog(@"rewardedAdDidFailToPresent");
  [self updateCallbackLogsWithTitle:@"rewardedAdDidFailToPresent"];
}

- (void)rewardedAdDidTrackImpression:(VungleRewarded *)rewarded {
  NSLog(@"rewardedAdDidTrackImpression");
  [self updateCallbackLogsWithTitle:@"rewardedAdDidTrackImpression"];
}

- (void)rewardedAdDidClick:(VungleRewarded *)rewarded {
  NSLog(@"rewardedAdDidClick");
  [self updateCallbackLogsWithTitle:@"rewardedAdDidClick"];
}

- (void)rewardedAdWillLeaveApplication:(VungleRewarded *)rewarded {
  NSLog(@"rewardedAdWillLeaveApplication");
  [self updateCallbackLogsWithTitle:@"rewardedAdWillLeaveApplication"];
}

- (void)rewardedAdDidRewardUser:(VungleRewarded *)rewarded {
  NSLog(@"rewardedAdDidRewardUser");
  [self updateCallbackLogsWithTitle:@"rewardedAdDidRewardUser"];
}

- (void)rewardedAdWillClose:(VungleRewarded *)rewarded {
  NSLog(@"rewardedAdWillClose");
  [self updateCallbackLogsWithTitle:@"rewardedAdWillClose"];
}

- (void)rewardedAdDidClose:(VungleRewarded *)rewarded {
  NSLog(@"rewardedAdDidClose");
  [self updateCallbackLogsWithTitle:@"rewardedAdDidClose"];
}

@end

File: Samples/iOS/Swift/IOS-SDK-Demo-App/Ad Formats/LORewardedViewController.swift
//
//  FullscreenViewController.swift
//  PublicSampleAppTrial
//
//  Created by John Mai on 12/29/22.
//

/**
 IMPORTANT: Please review and set the placementID and appID inside the Info.plist file. To do so, please do the following:
 
 Navigate to project hierarchy in Xcode, and select Info. Expand Liftoff PlacementID and be sure that Rewarded or Rewarded Playable is set correctly. Please do the same for Liftoff Dashboard AppID as well inside the Info.plist file.
 */

import UIKit
import VungleAdsSDK

class LORewardedViewController: UIViewController {
    private var rewardedAd: VungleRewarded?
    
    private var callbackLogs: [CallbackLog] = []
    var placementId:String?
    
    @IBOutlet weak var loadBtn: UIButton!
    @IBOutlet weak var playBtn: UIButton!
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var placementIdLbl: UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.navigationItem.title = "Liftoff Rewarded Ad"
        self.placementIdLbl.text = self.placementId
        self.callbackLogs = [
            CallbackLog(title: "rewardedAdDidLoad", index: 0),
            CallbackLog(title: "rewardedAdWillPresent", index: 1),
            CallbackLog(title: "rewardedAdDidPresent", index: 2),
            CallbackLog(title: "rewardedAdDidTrackImpression", index: 3),
            CallbackLog(title: "rewardedAdDidClick", index: 4),
            CallbackLog(title: "rewardedAdWillLeaveApplication", index: 5),
            CallbackLog(title: "rewardedAdDidRewardUser", index: 6),
            CallbackLog(title: "rewardedAdWillClose", index: 7),
            CallbackLog(title: "rewardedAdDidClose", index: 8),
            CallbackLog(title: "rewardedAdDidFailToLoad", index: 9),
            CallbackLog(title: "rewardedAdDidFailToPresent", index: 10)
        ]
        
        self.tableView.delegate = self
        self.tableView.dataSource = self
    }
    
    @IBAction func btnPressed(_ sender: UIButton) {
        if sender == self.loadBtn {
            // Clean up before load
            if self.rewardedAd != nil {
                self.rewardedAd?.delegate = nil
                self.rewardedAd = nil
                AppUtil.resetLogMessage(tableView: self.tableView, callbacks: self.callbackLogs)
            }
            
            guard let placementId = self.placementId else {
                return
            }
            self.rewardedAd = VungleRewarded(placementId: placementId)
            self.rewardedAd?.delegate = self
            self.rewardedAd?.load()
            
            // Set incentivized rewarded text
            self.rewardedAd?.setAlertTitleText("Close Ad")
            self.rewardedAd?.setAlertBodyText("Are you sure you want to close the ad? You will lose out on your reward")
            self.rewardedAd?.setAlertContinueButtonText("Resume Ad")
            self.rewardedAd?.setAlertCloseButtonText("Close Ad")
        } else if sender == self.playBtn {
            // Determines whether the banner object is nil in addition to whether the ad assets have been downloaded successfully
            if (self.rewardedAd?.canPlayAd() != nil) {
                self.rewardedAd?.present(with: self)
            }
        }
    }
}

// MARK: Vungle SDK Rewarded Delegate Callbacks

extension LORewardedViewController: VungleRewardedDelegate {
    // Ad load events
    func rewardedAdDidLoad(_ rewarded: VungleRewarded) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "rewardedAdDidLoad")
    }
    
    func rewardedAdDidFailToLoad(_ rewarded: VungleRewarded, withError: NSError) {
        print(#function)
        print(withError)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "rewardedAdDidFailToLoad")
    }
    
    // Ad Lifecycle Events
    func rewardedAdWillPresent(_ rewarded: VungleRewarded) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "rewardedAdWillPresent")
    }
    
    func rewardedAdDidPresent(_ rewarded: VungleRewarded) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "rewardedAdDidPresent")
    }
    
    func rewardedAdDidFailToPresent(_ rewarded: VungleRewarded, withError: NSError) {
        print(#function)
        print(withError)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "rewardedAdDidFailToPresent")
    }
    
    func rewardedAdDidTrackImpression(_ rewarded: VungleRewarded) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "rewardedAdDidTrackImpression")
    }
    
    func rewardedAdDidClick(_ rewarded: VungleRewarded) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "rewardedAdDidClick")
    }
    
    func rewardedAdWillLeaveApplication(_ rewarded: VungleRewarded) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "rewardedAdWillLeaveApplication")
    }
    
    func rewardedAdDidRewardUser(_ rewarded: VungleRewarded) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "rewardedAdDidRewardUser")
    }
    
    func rewardedAdWillClose(_ rewarded: VungleRewarded) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "rewardedAdWillClose")
    }
    
    func rewardedAdDidClose(_ rewarded: VungleRewarded) {
        print(#function)
        AppUtil.activeLogMessage(tableView: self.tableView, callbacks: self.callbackLogs, title: "rewardedAdDidClose")
    }
}

extension LORewardedViewController: UITableViewDelegate, UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.callbackLogs.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "CallbackLogTableCell",for: indexPath) as! CallbackLogTableViewCell
        let log = self.callbackLogs[indexPath.row]
        if log.isActive {
            cell.callbackLbl.textColor = .label
            cell.accessoryType = .checkmark
        } else {
            cell.callbackLbl.textColor = .gray
            cell.accessoryType = .none
        }
        cell.callbackLbl?.text = log.title
        return cell
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 60
    }
}
